bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/Journal.java

```diff
@@ -245,7 +246,8 @@ void readLog() {
                     }
                     bb.clear();
                     mark.readLogMark(bb);
-                    if (curMark.compare(mark) < 0) {
+                    // get the minimum mark position from all the ledger directories to ensure no data loss.
+                    if (curMark.compare(mark) > 0) {
```

Comment:
Since we are now replaying the journal from the smallest LogMark, can we get the journal log mark position of the current entry to the JournalScanner and compare it with the checkpoint position on the ledger disk where the entry to be restored is located, and only restore the entry whose logMark position is greater than the checkpoint? So as to avoid repeatedly writing the data that has been flushed to the disk.

Reply:
It will introduce complex logic for this comparison.
1. If the ledger directory expands or shrinks, the map logic of the ledgerId to the ledger directory (`logMark`) also changed. It may be located on the wrong `logMark` file, and will lead to skipping the unflushed entries.
2. There are many kinds of storage implementation, such as dbLedgerStorage, SortedLedgerStorage, and InterleavedLedgerStorage, we should get the ledgerId related storage instance to check the `logMark` position for each storage implementation. This operation will introduce complex logic.
3. For the comparison, we can only save the write ledger throughput. We also need to read the data from the journal log file out.

Based on the above reason, I prefer to replay all entries in the journal log file based on the min `logMark` position.